### PR TITLE
[Layer dialog] Prompt user to save unsaved changes when they click "Cancel"

### PR DESCRIPTION
**Pull request by parulraheja98**
In this PR , I only validate the form and check if the value in the form has changed. In a following PR I will check for remaining fields such as label, contributors may add etc.

Screen Shot 2021-08-10 at 9 24 07 PM

Contributes to #788.

### CODE REVIEW

**Comment by parulraheja98** [comment]
@gino-m , PTAL :)

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 1]
```diff
@@ -205,9 +205,24 @@ export class LayerDialogComponent implements OnDestroy {
       });
   }
 
-  onClose() {
-    this.dialogRef.close();
-    return this.navigationService.selectProject(this.projectId!);
+  onCancel(): void | undefined {
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:208]: Why `| undefined`? Can we remove this?

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 2]
```diff
@@ -276,4 +291,35 @@ export class LayerDialogComponent implements OnDestroy {
...
+    }
+    for (const editor of this.formFieldEditors) {
+      if (editor.formGroup.dirty || !this.isFieldOptionsDirty(editor)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private onClose(): void {
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:321]: `onClose()` => `close()`

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 3]
```diff
@@ -276,4 +291,35 @@ export class LayerDialogComponent implements OnDestroy {
...
+    this.dialogRef.close();
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:322]: This shouldn't be needed.

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 4]
```diff
@@ -276,4 +291,35 @@ export class LayerDialogComponent implements OnDestroy {
...
+    return this.navigationService.selectProject(this.projectId!);
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:323]: Please add TODO to replace this with a new method in `NavigationService`, something like `closeLayerDialog()`. This method just has to remove the `#l=` part of the URL.

**Review by gino-m — CHANGES_REQUESTED** [review]
A few discussed changes.

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 5]
```diff
@@ -205,9 +205,24 @@ export class LayerDialogComponent implements OnDestroy {
...
-    this.dialogRef.close();
-    return this.navigationService.selectProject(this.projectId!);
+  onCancel(): void | undefined {
+    if (!this.hasUnsavedChanges()) {
+      this.onClose();
+      return;
+    }
+    this.dialogService
+      .openConfirmationDialog(
+        'Warning',
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:215]: Please change to "Discard charges"?

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 6]
```diff
@@ -205,9 +205,24 @@ export class LayerDialogComponent implements OnDestroy {
...
+        'You have made unsaved changes to the layer. Would you like to save them?'
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:216]: "Unsaved changes to this layer will be lost. Are you sure?"

And set buttons to:
"CANCEL" and "DISCARD" as per https://material.io/components/dialogs#usage

Be sure to update the conditions in the subscribe callback below as well.

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 7]
```diff
@@ -205,9 +205,24 @@ export class LayerDialogComponent implements OnDestroy {
...
+      )
+      .afterClosed()
+      .subscribe(async dialogResult => {
+        if (dialogResult) {
+          await this.onSave();
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:221]: Replace this call with `this.close()` here.

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts [hunk 8]
```diff
@@ -205,9 +205,24 @@ export class LayerDialogComponent implements OnDestroy {
...
+        } else {
```
Comment (gino-m) [inline web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:222]: And remove this branch.

**Review by gino-m — CHANGES_REQUESTED** [review]

**Comment by parulraheja98** [comment]
Screen Shot 2021-08-31 at 9 27 35 PM

File: web-ng/src/app/components/layer-dialog/layer-dialog.component.ts (hunk 3 above)
Comment (parulraheja98) [reply web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:321]: I tried removing this but the dialog doesn't close.

Comment (gino-m) [reply web-ng/src/app/components/layer-dialog/layer-dialog.component.ts:321]: Sorry for my ignorance here! Please disregard.

**Review by gino-m — APPROVED** [review]
Nice work!

**Pull request merged**